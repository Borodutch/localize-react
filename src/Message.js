import { useLocalize } from './use-localize';

export function Message({ descriptor }) {
  const { translate } = useLocalize();

  return translate(descriptor);
}
