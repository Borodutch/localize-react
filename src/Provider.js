import React from 'react';
import { sanitizeLocale, memoize } from './helpers';

export const LocalizationContext = React.createContext();

export function LocalizationProvider({ children, locale, translations = {} }) {
  const sanitizedLocale = sanitizeLocale(locale, translations);
  const pureTranslations = sanitizedLocale ? translations[sanitizedLocale] : translations;

  function pureTranslateFn(key) {
    if (!pureTranslations || !key) return key;
    if (typeof pureTranslations[key] === 'string') return pureTranslations[key];

    const complexKeyArray = key.split('.');
    if (complexKeyArray.length === 1) return key;

    let finalValue = pureTranslations[complexKeyArray[0]];
    for (let i = 1; i < complexKeyArray.length; i++) {
      if (finalValue) {
        finalValue = finalValue[complexKeyArray[i]];
      }
    }

    return typeof finalValue === 'string' ? finalValue : key;
  }

  const translate = memoize(pureTranslateFn);

  return (
    <LocalizationContext.Provider value={{ locale, translate, translations }}>
      {children}
    </LocalizationContext.Provider>
  );
}

export const LocalizationConsumer = LocalizationContext.Consumer;
