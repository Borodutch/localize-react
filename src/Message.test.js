import React from 'react';
import renderer from 'react-test-renderer';
import { LocalizationProvider } from './Provider';
import { Message } from './Message';

const TRANSLATIONS = {
  en: {
    name: 'Alex',
  },
};

describe('Message', () => {
  it('translates from key to text', () => {
    const tree = renderer
      .create(
        <LocalizationProvider
          locale="en"
          translations={TRANSLATIONS}
        >
          <Message descriptor="name" />
        </LocalizationProvider>
      )
      .toJSON();

      expect(tree).toMatchSnapshot();
  });
});
