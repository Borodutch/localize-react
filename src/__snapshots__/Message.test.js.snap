// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message translates from key to text 1`] = `"Alex"`;

exports[`Message translates from key to text with default message 1`] = `"Default Message"`;
